### ISSUE TITLE
Aggregating (count distinct) without grouping results in single segment doing all the work

### ISSUE THREAD

**Issue by andrea-s** [greenplum-db/gpdb#4924]
TLDR: a count distinct with no grouping produces a very inefficient query plan when using GPORCA - even grouping by a dummy value produces a more sensible query plan.

# 1) without grouping
explain select count(distinct visitor_id) from mydb.mytable;
                                                     QUERY PLAN
 Aggregate  (cost=0.00..431.00 rows=1 width=8)
   ->  Gather Motion 64:1  (slice1; segments: 64)  (cost=0.00..431.00 rows=1 width=8)
         ->  Sequence  (cost=0.00..431.00 rows=1 width=8)
               ->  Partition Selector for mytable (dynamic scan id: 1)  (cost=10.00..100.00 rows=2 width=4)
                     Partitions selected: 33 (out of 33)
               ->  Dynamic Table Scan on mytable (dynamic scan id: 1)  (cost=0.00..431.00 rows=1 width=8)
 Optimizer status: PQO version 2.55.13

When executed, it looks like a single segment is dong *all* of the work here and because of that the query takes a long time.

# 2) with dummy grouping
explain select 'someconst' as const, count(distinct visitor_id) from mydb.mytable group by const;
                                                                       QUERY PLAN
 Gather Motion 64:1  (slice2; segments: 64)  (cost=0.00..431.00 rows=1 width=16)
   ->  GroupAggregate  (cost=0.00..431.00 rows=1 width=16)
         Group By: const
         ->  Sort  (cost=0.00..431.00 rows=1 width=16)
               Sort Key: const
               ->  Redistribute Motion 64:64  (slice1; segments: 64)  (cost=0.00..431.00 rows=1 width=16)
                     Hash Key: const
                     ->  Result  (cost=0.00..431.00 rows=1 width=16)
                           ->  GroupAggregate  (cost=0.00..431.00 rows=1 width=16)
                                 Group By: const
                                 ->  Sort  (cost=0.00..431.00 rows=1 width=16)
                                       Sort Key: const, visitor_id
                                       ->  Result  (cost=0.00..431.00 rows=1 width=16)
                                             ->  Sequence  (cost=0.00..431.00 rows=1 width=8)
                                                   ->  Partition Selector for mytable (dynamic scan id: 1)  (cost=10.00..100.00 rows=2 width=4)
                                                         Partitions selected: 33 (out of 33)
                                                   ->  Dynamic Table Scan on mytable (dynamic scan id: 1)  (cost=0.00..431.00 rows=1 width=8)
 Optimizer status: PQO version 2.55.13

This query, which is functionally equivalent, gets correctly distributed over the segments and is much faster as a result.

### Greenplum version or build
PostgreSQL 8.3.23 (Greenplum Database 5.7.0 build f7c6eb5-oss) on x86_64-pc-linux-gnu, compiled by GCC gcc (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609, 64-bit compiled on Mar 30 2018 19:36:48

### OS version and uname -a
Linux Ubuntu-1604-xenial-64-minimal 4.13.0-36-generic #40~16.04.1-Ubuntu SMP Fri Feb 16 23:25:58 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

### Installation information ( pg_config )
BINDIR = /opt/gpdb/bin
DOCDIR = /opt/gpdb/doc/postgresql
HTMLDIR = /opt/gpdb/doc/postgresql
INCLUDEDIR = /opt/gpdb/include
PKGINCLUDEDIR = /opt/gpdb/include/postgresql
INCLUDEDIR-SERVER = /opt/gpdb/include/postgresql/server
LIBDIR = /opt/gpdb/lib
PKGLIBDIR = /opt/gpdb/lib/postgresql
LOCALEDIR = /opt/gpdb/share/locale
MANDIR = /opt/gpdb/share/man
SHAREDIR = /opt/gpdb/share/postgresql
SYSCONFDIR = /opt/gpdb/etc/postgresql
PGXS = /opt/gpdb/lib/postgresql/pgxs/src/makefiles/pgxs.mk
CC = gcc
CPPFLAGS = -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/opt/gpdb/include
CFLAGS = -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -std=gnu99 -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -I/opt/gpdb/include
CFLAGS_SL = -fpic
LDFLAGS = -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,--as-needed -Wl,-rpath,'/build/greenplum-db-oss-feJIiP/greenplum-db-oss-5.7.0/debian/build/lib',--enable-new-dtags -L/opt/gpdb/lib
LDFLAGS_SL =
LIBS = -lpgport -lgpopt -lnaucrates -lgpdbcost -lgpos -lxerces-c -lxml2 -lrt -lgssapi_krb5 -lcrypt -ldl -lm -L/opt/gpdb/lib
VERSION = PostgreSQL 8.3.23

**Comment by vraghavan78** [comment]
@andrea-s currently GPORCA is forced to pick two stage aggregate when the grouping column is not the distribution column. (This is because the GUC optimizer_force_multistage_agg is set to true). Given that the table has no data, and turning optimizer_force_multistage_agg to off will generate a single stage aggregate with all the data passed to the master.

Reasoning why the GUC is set to on is as follows: Most customers / users of GPDB have large scale data in TB or PB on big clusters. The aggregates are usually done on top of large subqueries where all cardinality estimation is out of the window. This GUC prevents cardinality misestimation to skew the cost in this common used usecase.

Currently GPORCA does not consider constant to be similar to distribution column. It is good suggestion that we should do. That being said, grouping on constant while a valid SQL syntax is not widely used in my observation.

**Comment by d** [comment]
@andrea-s your table has no data. So your issue is a non-issue
